internal/datanode/event_manager.go

```diff
@@ -0,0 +1,101 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datanode
+
+import (
+	"time"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+
+	"go.uber.org/zap"
+)
+
+const retryWatchInterval = 20 * time.Second
+
+type event struct {
+	eventType int
+	vChanName string
+	info      *datapb.ChannelWatchInfo
+}
+
+type channelEventManager struct {
+	eventChan         chan event
+	closeChan         chan struct{}
+	handlePutEvent    func(watchInfo *datapb.ChannelWatchInfo) error // node.handlePutEvent
+	handleDeleteEvent func(vChanName string)                         // node.handleDeleteEvent
+}
+
+const (
+	putEventType    = 1
+	deleteEventType = 2
+)
+
+func (e *channelEventManager) Run() {
```

Comment:
better to add sync.waitGroup and wait in Close().
better to extract the main logic into a separate function.